### PR TITLE
introduce libxsmm

### PR DESCRIPTION
**Pull request by tensor-tang**
The benchmark data shown at [this](https://github.com/tensor-tang/libxsmm/blob/master/README.md#benchmark)

### CODE REVIEW

**Comment by tensor-tang** [comment]
https://paddleci.ngrok.io/viewLog.html?tab=buildLog&logTab=tree&filter=debug&expand=all&buildId=1292&_focus=6439
```
[Step 1/1] In file included from /paddle/paddle/fluid/framework/tensor.h:26:0,
[03:10:52][Step 1/1]                  from /paddle/paddle/fluid/framework/mixed_vector.h:21,
[03:10:52][Step 1/1]                  from /paddle/paddle/fluid/framework/lod_tensor.h:28,
[03:10:52][Step 1/1]                  from /paddle/paddle/contrib/inference/paddle_inference_api_impl.h:25,
[03:10:52][Step 1/1]                  from /paddle/paddle/contrib/inference/paddle_inference_api_impl.cc:24:
[03:10:52][Step 1/1] /paddle/paddle/fluid/platform/device_context.h:27:22: fatal error: mkldnn.hpp: No such file or directory
```

**Comment by tensor-tang** [comment]
测试了paddle/fluid/inference/tests/book/test_inference_nlp
发现速度有drop,原因在于mnk的阈值选择，不能参考github的官网推荐阈值。

Items           | Before  | After
:--------------- | :------ | :------
latency @1 core  | 2.26    | 2.63
QPS @1 core      | 441.86  | 380.49
latency @10 core | 3.42    | 3.61
QPS @10 core     | 2647.72 | 2509.76

**Comment by tensor-tang** [comment]
Items           | Before  | After | New
:--------------- | :------ | :------ | :------
latency @1 core  | 2.26    | 2.63 | 2.33
QPS @1 core      | 441.86  | 380.49 | 433
latency @10 core | 3.42    | 3.61 | 3.45
QPS @10 core     | 2647.72 | 2509.76|2609

改了阈值后速度恢复到之前水平。

比之前的workload稍微慢一点原因是Before的develop还早几个commit，发现benchmark的nightly很需要。

**Comment by tensor-tang** [comment]
https://paddleci.ngrok.io/viewLog.html?tab=buildLog&logTab=tree&filter=debug&expand=all&buildId=1544&_focus=10816
```
[Step 1/1] API Difference is: 
[13:53:54][Step 1/1] - paddle.fluid.layers.im2sequence ArgSpec(args=['input', 'filter_size', 'stride', 'padding', 'name'], varargs=None, keywords=None, defaults=(1, 1, 0, None))
[13:53:54][Step 1/1] + paddle.fluid.layers.im2sequence ArgSpec(args=['input', 'filter_size', 'stride', 'padding', 'input_image_size', 'out_stride', 'name'], varargs=None, keywords=None, defaults=(1, 1, 0, None, 1, None))
[13:53:54][Step 1/1] ?     
```

File: paddle/fluid/operators/math/blas_impl.h [hunk 1]
```diff
@@ -140,13 +153,43 @@ struct CBlas<double> {
...
 };
 
+template <typename T>
+inline bool UseXSMM(const int &m, const int &n, const int &k, bool transa,
+                    bool transb, const T &alpha, const T &beta) {
+#ifdef PADDLE_WITH_LIBXSMM
+  // Refer to https://github.com/hfp/libxsmm/blob/master/README.md
+  // But the threshold is custom
+  constexpr int LIBXSMM_THRESHOLD = 20 * 20 * 20;
+  if (m * n * k > LIBXSMM_THRESHOLD || transa || transb ||
```
Comment (Superjomn) [inline paddle/fluid/operators/math/blas_impl.h:173]: Might better to enable this to be changed in the runtime latter to tune the performance on different CPU type or other environments.

Currently, this implementation is OK.

File: paddle/fluid/operators/math/math_function_test.cc [hunk 2]
```diff
@@ -54,8 +54,64 @@ TEST(math_function, gemm_notrans_cblas) {
...
+  };
+
+  smm();
+  mkl();
+  ASSERT_EQ(mat_c_mkl.numel(), mat_c_smm.numel());
+  for (int i = 0; i < mat_c_mkl.numel(); ++i) {
+    EXPECT_FLOAT_EQ(CSMM[i], CMKL[i]);
+  }
+}
+TEST(math_function, gemm_mkl_vs_smm) {
```
Comment (Superjomn) [inline paddle/fluid/operators/math/math_function_test.cc:104]: Leave an empty line above.

File: paddle/fluid/operators/math/math_function_test.cc [hunk 3]
```diff
@@ -54,8 +54,64 @@ TEST(math_function, gemm_notrans_cblas) {
   EXPECT_EQ(input3_ptr[6], 86);
   EXPECT_EQ(input3_ptr[7], 99);
 }
+#ifdef PADDLE_WITH_LIBXSMM
```
Comment (Superjomn) [inline paddle/fluid/operators/math/math_function_test.cc:57]: Leave an empty line above.

File: paddle/fluid/operators/math/blas_impl.h (hunk 1 above)
Comment (tensor-tang) [reply paddle/fluid/operators/math/blas_impl.h:173]: Yes, I will refine this in next PR.

File: paddle/fluid/operators/math/math_function_test.cc (hunk 3 above)
Comment (tensor-tang) [reply paddle/fluid/operators/math/math_function_test.cc:57]: Yes, I will update in next PR

**Review by luotao1 — APPROVED** [review]
LGTM！

**Comment by luotao1** [comment]
WITH_LIBXSMM 这个开关默认是关的，因此在ci上无法验证，之后加入CE来验证吧。

**Comment by tensor-tang** [comment]
嗯，以后肯定是需要验证的。

**Pull request merged**